src/components/utils/PreviewedLink.tsx

```diff
@@ -0,0 +1,149 @@
+
+import SafeLink from '@/components/utils/SafeLink';
+import { isTwitterUrl, isYoutubeVideoUrl } from '@/utils/url';
+
+import '@/types/twitterWidget.d';
+
+type SafeLinkProps = {
+  class?: string;
+  href: string;
+  children?: JSX.Element;
+};
+
+type OgpContents = {
+  url: URL | null;
+  title: string;
+  description: string;
+  image: URL | null;
+};
+
+const twitterUrl = (urlString: string): string => {
+  try {
+    const url = new URL(urlString);
+    url.host = 'twitter.com';
+    return url.href;
+  } catch {
+    return '';
+  }
+};
+
+const youtubeUrl = (urlString: string): string => {
+  try {
+    const originalUrl = new URL(urlString);
+    const iframeUrl = new URL('https://www.youtube.com/embed/');
+    iframeUrl.pathname += originalUrl.searchParams.get('v');
+    iframeUrl.searchParams.set('origin', window.location.href);
+    return iframeUrl.href;
+  } catch {
+    return '';
+  }
+};
+
+const fetchOgpContents = async (url: URL): Promise<OgpContents | null> => {
+  const whiteList = ['www3.nhk.or.jp'];
+  if (whiteList.includes(url.host)) {
+    const res = await fetch(url, { headers: { Accept: 'text/html' } });
+    const text = await res.text();
+    const el = new DOMParser().parseFromString(text, 'text/html');
+    const ogs: { [index: string]: string } = {};
+    Array.from(el.head.querySelectorAll('meta'))
+      .filter((v) => v.getAttribute('property') != null && v.getAttribute('content') != null)
+      .forEach((v) => {
+        const property = v.getAttribute('property');
+        const content = v.getAttribute('content');
+        if (property != null && content != null) {
+          ogs[property] = content;
+        }
+      });
+
+    if (ogs['og:image'] && ogs['og:title'] && ogs['og:description']) {
+      const contents: OgpContents = {
+        title: ogs['og:title'],
+        description: ogs['og:description'],
+        image: new URL(ogs['og:image']),
+        url,
+      };
+      return contents;
+    }
+  }
+  return null;
+};
+
+const PreviewedLink: Component<SafeLinkProps> = (props) => {
+  const isSafe = () => {
+    try {
+      const url = new URL(props.href.toString());
+      return url.protocol === 'https:' || url.protocol === 'http:';
+    } catch {
+      return false;
+    }
+  };
+
+  let twitterRef: HTMLQuoteElement | undefined;
+
+  createEffect(() => {
+    if (isTwitterUrl(props.href.toString())) {
+      window.twttr?.widgets?.load(twitterRef);
+    }
+  });
+  const initialOgp: OgpContents = { title: '', description: '', image: null, url: null };
+  const [getOgpContents, setOgpContents] = createSignal(initialOgp);
+
+  const updateOgpContents = async () => {
+    const ogp = await fetchOgpContents(new URL(props.href.toString()));
+    if (ogp != null) {
+      setOgpContents(ogp);
+    }
+  };
+  onMount(() => {
+    updateOgpContents()
+      .then(() => {})
+      .catch(() => {});
+  });
```

Comment:
Reactの習慣に倣って、`ref`→状態関連→その他の関数→createEffect→onMountの順にしていただけると嬉しいです:

```suggestion
  let twitterRef: HTMLQuoteElement | undefined;

  const initialOgp: OgpContents = { title: '', description: '', image: null, url: null };
  const [getOgpContents, setOgpContents] = createSignal(initialOgp);

  const isSafe = () => {
    try {
      const url = new URL(props.href.toString());
      return url.protocol === 'https:' || url.protocol === 'http:';
    } catch {
      return false;
    }
  };

  const updateOgpContents = async () => {
    const ogp = await fetchOgpContents(new URL(props.href.toString()));
    if (ogp != null) {
      setOgpContents(ogp);
    }
  };
  
  createEffect(() => {
    if (isTwitterUrl(props.href.toString())) {
      window.twttr?.widgets?.load(twitterRef);
    }
  });
  
  onMount(() => {
    updateOgpContents()
      .then(() => {})
      .catch(() => {});
  });
```